wgpu-hal/src/gles/queue.rs

```diff
@@ -25,17 +27,112 @@ fn is_3d_target(target: super::BindTarget) -> bool {
     }
 }
 
+/// Helper macro to define repetative `set_[something]_enabled` functions in Queue impl
+macro_rules! gl_modes_set_enabled_fn {
+    ($fn_name:ident, $var_name:ident, $gl_var:ident) => {
+        unsafe fn $fn_name(&self, enabled: bool) {
+            self.state.lock().unwrap().$var_name = enabled;
+            if enabled {
+                self.shared.context.enable(glow::$gl_var);
+            } else {
+                self.shared.context.disable(glow::$gl_var);
+            }
+        }
+    };
+}
+
 impl super::Queue {
-    unsafe fn reset_state(&self) {
+    // Add helper functions for enabling and disabling certain GL features so that we can keep track
+    // of the current enabled/disabled state
+    unsafe fn set_current_program(&self, program: Option<glow::Program>) {
+        self.shared.context.use_program(program);
+        self.state.lock().unwrap().current_program = program;
+    }
+    gl_modes_set_enabled_fn!(
```

Comment:
Are these changes necessary for the purpose of the PR? I think they aren't necessary.
Let's please focus on the goal here, otherwise it's much more difficult to land.